[WIP] [editor] Renaming a file should rename a corresponding editor

lmcgupe opened the pull request:
Signed-off-by: Guy Perron <guy.perron@ericsson.com>

hexa00 commented:
I tought you were trying to close and open just for experimentation.

But if you ask for review, you should know this is not the proper way forward.
You should listen on filesystem changes and update the title for the widget.

svenefftinge commented:
We don't need anything in the widget-manager for closing. Just call `widget.close()`.

lmcgupe commented:
The scope of this fix is when the rename is done within the IDE.

hexa00 commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -90,6 +90,21 @@ export class WidgetManager {
+        let key = this.toKey({ factoryId, options });
+        let existingWidget = this.widgets.get(key);
+        if (existingWidget) {
+            existingWidget.title.label = newLabel;
```
Changing the label of a widget and changing it's key are 2 different things.

You should have a method to change the key here called updateWidgetKey for example.

And the title should be changed after filesysem.move is done and sucessful

akosyakov replied:
It should happen in the editor manager, nothing to do with the widget manager

hexa00 commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -104,9 +104,11 @@ export class WorkspaceCommandContribution implements CommandContribution {
+                dialog.open().then(name => {
+                    renameWidget(this.openerService, uri.parent.resolve(uri.path.base), uri.parent.resolve(name), name).then(() => {
+                        this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString());
+                    });
```
we should:
  - move the file and make sure it worked
 -  update the widget key
 - change the title

akosyakov commented on packages/core/src/browser/opener-service.ts:
```diff
@@ -43,6 +43,8 @@ export interface OpenHandler {
      */
     open(uri: URI, options?: OpenerOptions): MaybePromise<object | undefined>;
+
+    renameWidget(uri: URI, newUri: URI, newLabel: string, options?: OpenerOptions): void;
```
You should not change `OpenerService`, only  `EditorManager` should be changed.

akosyakov commented:
To make it right, i think it shold be based on [Resource](https://github.com/theia-ide/theia/blob/05f2d772f7c6effdcc9afb2642944b5786d4ba6a/packages/core/src/common/resource.ts#L15). Editors are not working with files, but with resources, e.g. java resource to open source code, git resource to read history and so on.

@svenefftinge What do you think about adding optional `onDidChangeUri` event to `Resource` interface. `FileResource` it would listen to file changes and if resource is renamed when notify an editor that it could update label and icon.

akosyakov commented:
To implement it we would need also expose an underlying editor resource via `TextEditor` interface to allow `EditorManager` to access it here: 
https://github.com/theia-ide/theia/blob/05f2d772f7c6effdcc9afb2642944b5786d4ba6a/packages/editor/src/browser/editor-manager.ts#L116-L121

So it would be here something like:
```ts
if (newEditor.resource.onDidChangeUri) {
  newEditor.resource.onDidChangeUri(newUri => {
    newEditor.title.label = newUri.path.base; 
    newEditor.title.iconClass = this.iconProvider.getFileIconForURI(textEditor.newUri);
  });
}
```

akosyakov commented:
We also don't have an event so far to indicate a file rename, you should look in the chokidar watcher code in vscode to learn how to identify it. I saw something like that there.

hexa00 commented:
Nice thanks for the pointers @akosyakov! 

Note I think we could do this in 2 steps, first use the UI rename as a trigger to the onDidChangeUri and after use the FileWatcher.

Also note that vscode doesn't rename if a file is moved from the filesystem, it marks the editor as delete from disk if you had unsaved changes or it closes the editor if you had no change.

svenefftinge commented:
Modifying a widget after creation will break the contract of the widget manager. After reload it will try to create a widget based on the old uri.

akosyakov commented:
> Modifying a widget after creation will break the contract of the widget manager. After reload it will try to create a widget based on the old uri.

Won't it flicker and lose an editor content if we just close and open another?

akosyakov commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -90,6 +90,28 @@ export class WidgetManager {
+     *  check if a widget exists already and rename it
+     */
+    renameWidget(factoryId: string, options: any, newOptions: any, newLabel: any): void {
+        let key = this.toKey({ factoryId, options });
```
please check that tslint is enabled in VS code, you should have warnings here

akosyakov commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -61,7 +62,9 @@ export class WorkspaceCommandContribution implements CommandContribution {
         @inject(SelectionService) protected readonly selectionService: SelectionService,
-        @inject(OpenerService) protected readonly openerService: OpenerService
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(EditorManagerImpl) private editorService: EditorManagerImpl
```
program against the interface not implementation, i.e. `EditorManager`

lmcgupe replied:
I moved the code over to the EditorManager, this introduced a dependency on the editor package. But now since the editor package also has a dependency on workspace, it won't compile, complaining that editor package doesn't exist (circular dependency). Any idea ?

akosyakov replied:
Could you try to remove a dependency from `@theia/editor` to `@theia/workspace`? The latter does not seem to be used in the former.

akosyakov commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -104,9 +107,11 @@ export class WorkspaceCommandContribution implements CommandContribution {
-                    this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString())
-                );
+                dialog.open().then(name => {
+                    this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString()).then(() => {
```
I think @svenefftinge meant that you only should close an existing editor and open new:
```ts
const newUri = uri.parent.resolve(name);
this.fileSystem.move(uri.toString(), newUri.toString()).then(() => {
  const widget = this.editorService.getEditor(uri);
  if (editor) {
    const selection = widget.editor.selection;
    widget.close();
    open(this.openerService, newUri, {selection});
  }
});
```

Ideally, it should be done not here but on move/rename events from the filesystem. Making it here:
- breaks the unidirectional flow, only the filesystem should be the source of truth;
- cover only a specific case, e.g. renaming files in Finder won't make the same effect;
- make the workspace extension depend on the editor extension.

svenefftinge replied:
No, I did not mean that a real rename would be better. But we have to take care of all the loose ends.

hexa00 replied:
Maybe we should have an generic key to reference the widget directly to avoid lose ends ?

svenefftinge replied:
Could you elaborate on what you mean?

hexa00 replied:
I mean that if something has to keep a reference to an editor widget it should reference the widget object directly or we should provide a key that is not bound to any resource to refer to it. Like some abstract id.

So the URI would be used as key to open a new widget or find a widget for that URI but once you get a reference to it, it should be the id or the object directly, a bit like files names and files descriptors.. you open(filename) get a file descriptor so you could getWidgetForURI() get a widget for a URI but use getWidgetId() if you already had a widget and now know only it's id. I don't see where however that would be useful.. just keeping the object seems ok since serializing that doesn't make sense as the widget-manager is in the frontend.

>Modifying a widget after creation will break the contract of the widget manager. After reload it will try to create a widget based on the old uri.

Why would it try to reload with an old uri ? if you updated it before you serialized the widget ?

Granted if it's changed between save and restore there's nothing we can do but that is true whatever we do...

svenefftinge commented:
> Won't it flicker, lose an editor context (selection, viewport and so on) and break an existing layout if we just close one and open another?
Yes, that's probably not a good solution, but simply changing the widget state doesn't work either.

lmcgupe commented:
hum.. with the workspace dependency removed from editor, now it makes another error pop up. In the editor package it complains that preferences is not there.

 [compile] src/browser/editor-preferences.ts(16,8): error TS2307: Cannot find module '@theia/preferences/lib/common


Is there a way to specify the order of the compilation ?




________________________________
From: Anton Kosyakov <notifications@github.com>
Sent: October 2, 2017 10:32:04 AM
To: theia-ide/theia
Cc: Guy Perron; Author
Subject: Re: [theia-ide/theia] [WIP] [editor] Renaming a file should rename a corresponding editor (#582)


@akosyakov commented on this pull request.

________________________________

In packages/workspace/src/browser/workspace-commands.ts<https://github.com/theia-ide/theia/pull/582#discussion_r142155363>:

> @@ -61,7 +62,9 @@ export class WorkspaceCommandContribution implements CommandContribution {
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(WorkspaceServer) protected readonly workspaceServer: WorkspaceServer,
         @inject(SelectionService) protected readonly selectionService: SelectionService,
-        @inject(OpenerService) protected readonly openerService: OpenerService
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(EditorManagerImpl) private editorService: EditorManagerImpl


Could you try to remove a dependency from @theia/editor to @theia/workspace? The latter does not seem to be used in the former.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/theia-ide/theia/pull/582#discussion_r142155363>, or mute the thread<https://github.com/notifications/unsubscribe-auth/Aa23i_sYEAdOqprk6VeEt5qnn8nJ18ZFks5soPPkgaJpZM4PmiDZ>.

lmcgupe commented:
It actually fails now because preferences has a dependency on workspace. So we are back again with another circular dependency : workspace - editor - preferences.
That's since the introduction of the editor dependency in workspace.

You can fetch my patch if you want to try it out.


________________________________
From: Guy Perron
Sent: October 2, 2017 2:04:55 PM
To: theia-ide/theia
Subject: Re: [theia-ide/theia] [WIP] [editor] Renaming a file should rename a corresponding editor (#582)


hum.. with the workspace dependency removed from editor, now it makes another error pop up. In the editor package it complains that preferences is not there.

 [compile] src/browser/editor-preferences.ts(16,8): error TS2307: Cannot find module '@theia/preferences/lib/common


Is there a way to specify the order of the compilation ?




________________________________
From: Anton Kosyakov <notifications@github.com>
Sent: October 2, 2017 10:32:04 AM
To: theia-ide/theia
Cc: Guy Perron; Author
Subject: Re: [theia-ide/theia] [WIP] [editor] Renaming a file should rename a corresponding editor (#582)


@akosyakov commented on this pull request.

________________________________

In packages/workspace/src/browser/workspace-commands.ts<https://github.com/theia-ide/theia/pull/582#discussion_r142155363>:

> @@ -61,7 +62,9 @@ export class WorkspaceCommandContribution implements CommandContribution {
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(WorkspaceServer) protected readonly workspaceServer: WorkspaceServer,
         @inject(SelectionService) protected readonly selectionService: SelectionService,
-        @inject(OpenerService) protected readonly openerService: OpenerService
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(EditorManagerImpl) private editorService: EditorManagerImpl


Could you try to remove a dependency from @theia/editor to @theia/workspace? The latter does not seem to be used in the former.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/theia-ide/theia/pull/582#discussion_r142155363>, or mute the thread<https://github.com/notifications/unsubscribe-auth/Aa23i_sYEAdOqprk6VeEt5qnn8nJ18ZFks5soPPkgaJpZM4PmiDZ>.

akosyakov commented:
`@theia/editor` should not depend on `@theia/preferences` only on `@theia/preferences-api`

lmcgupe commented:
Changing to preferences-api got one step further. Now it fails to find 
[compile] src/browser/editor-manager.ts(16,34): error TS2307: Cannot find module '@theia/filesystem/lib/browser/icons/file-icons'.

lmcgupe commented:
got over it after adding filesystem dependency inside editor's package.json

lmcgupe commented:
> To implement it we would need also expose an underlying editor resource via TextEditor interface to allow EditorManager to access it here:
I modified common/resource.ts by add an onDidChangeUri. Added a resource field inside browser/editor.ts. 
In a similar way what is done with OnDidChangeUriEmitter. Concerning this latter method, I tried hitting a break point inside file-resource.ts, in the onDidChangeContents method, after modifying a file, but it didn't trigger. Is that supposed to work ?

akosyakov commented:
You should put a breakpoint here: https://github.com/theia-ide/theia/blob/31d18c38aa08e6b81d127fdc8de3a03d6796dbbf/packages/filesystem/src/common/file-resource.ts#L28

`onDidChangeContents` is only hit when someone wants to install a listener, we don't have such clients in the code base, but I used it in the markdown extension:
https://github.com/theia-ide/theia-extension-example/blob/0731d92b3ede2d46e60ddaf61c1e6ff526ad4ad4/theia-markdown/src/browser/markdown-resource.ts#L33-L35

epatpol commented:
@akosyakov I didn't find anything for a file "rename" in chokidar i.e https://github.com/paulmillr/chokidar/issues/303. I guess we would have to implement it chokidar-client side ourselves?

epatpol commented:
> Note I think we could do this in 2 steps, first use the UI rename as a trigger to the onDidChangeUri and after use the FileWatcher.

I'm not sure about doing it in 2 steps though, how would the ui trigger the onDidChangeUri method of file-resource? Would we add a temporary method to that resource that can trigger the event from outside (instead of from the fs-watcher). Sorry maybe I misunderstood but I'm having an easier time understanding the event from the FS in the backend then UI -> Resource.OnDidChangeUri.fire() (usually the emitter is private right?).

akosyakov commented:
I meant to look into how vscode using chokidar. I've just looked myself, I was wrong: they are not able to detect rename as well and it only works if it triggered via UI.

hexa00 commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -104,9 +108,16 @@ export class WorkspaceCommandContribution implements CommandContribution {
+                    this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString()).then(() => {
+                        this.editorService.renameWidget(uri.parent.resolve(uri.path.base), uri.parent.resolve(name), name);
+                        let event: IUriChangedEvent = { factoryId: uri.parent.resolve(uri.path.base), options: uri.parent.resolve(name), newOptions: name, newLabel: ' ' };
+                        this.uiWatcher.onIUriChangedEmitter.fire(event);
```
couple of things
 - you don't need a watcher
 - nothing actually uses that event at the moment ?
 - you still call editorservice rename-widget

It should be that workspace-commands fires the event
and editor service receive it at the very least

Might be other considerations too.. need to think about this more

hexa00 commented on packages/workspace/src/browser/workspace-frontend-module.ts:
```diff
@@ -36,8 +36,11 @@ export default new ContainerModule((bind: interfaces.Bind, unbind: interfaces.Un
         (props: FileDialogProps) =>
             createFileDialog(ctx.container, props)
     );
+    bind(WorkspaceCommandContribution).toSelf().inSingletonScope();
```
This should not be there

hexa00 commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -56,7 +56,8 @@ export class WidgetManager {
         @inject(ContributionProvider) @named(WidgetFactory) protected readonly factoryProvider: ContributionProvider<WidgetFactory>,
-        @inject(ILogger) protected logger: ILogger) {
+        @inject(ILogger) protected logger: ILogger
+    ) {
```
no need to change this

hexa00 commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -69,6 +70,10 @@ export class WidgetManager {
+    getWidgetsMap(): Map<string, Widget> {
+        return this.widgets;
+    }
+
```
this is wrong you should not expose the internal map

hexa00 commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -105,7 +110,7 @@ export class WidgetManager {
     }
 
-    protected toKey(options: WidgetConstructionOptions) {
+    public toKey(options: WidgetConstructionOptions) {
```
You should not need this public either

hexa00 commented on packages/editor/src/browser/editor-frontend-module.ts:
```diff
@@ -16,6 +16,7 @@ import { EditorCommandHandlers } from "./editor-command";
 import { EditorKeybindingContribution, EditorKeybindingContext } from "./editor-keybinding";
 import { bindEditorPreferences } from './editor-preferences';
 import { WidgetFactory } from '@theia/core/lib/browser/widget-manager';
+// import { WorkspaceCommandContribution } from '@theia/workspace/lib/browser/workspace-commands';
```
remove

hexa00 commented on packages/editor/src/browser/editor-frontend-module.ts:
```diff
@@ -30,4 +31,5 @@ export default new ContainerModule(bind => {
     bind(EditorKeybindingContext).toSelf().inSingletonScope();
     bind(KeybindingContext).toDynamicValue(context => context.container.get(EditorKeybindingContext));
     bind(KeybindingContribution).to(EditorKeybindingContribution);
+    // bind(WorkspaceCommandContribution).toSelf().inSingletonScope();
```
remove

hexa00 commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -27,6 +29,7 @@ export interface EditorManager extends OpenHandler {
      * Reject if the given input is not an editor input or an editor cannot be opened.
      */
     open(uri: URI, input?: EditorInput): Promise<EditorWidget>;
+
```
remove

hexa00 commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -105,6 +112,28 @@ export class EditorManagerImpl implements EditorManager, WidgetFactory {
+        if (existingWidget) {
+            existingWidget.title.label = newLabel;
+            this.updateWidgetKey(key, existingWidget, factoryId, options, newOptions);
+        }
```
This is just externalizing your original implementation into the widget manager.
You should use EditorManager methosd to get to your widget.

hexa00 commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -105,6 +112,28 @@ export class EditorManagerImpl implements EditorManager, WidgetFactory {
+        options = newOptions;
+        key = this.widgetManager.toKey({ factoryId, options });
+        this.widgetManager.getWidgetsMap().set(key, widget);
+    }
```
Ditto... this one makes sense in the widget-manager

hexa00 commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -104,9 +130,17 @@ export class WorkspaceCommandContribution implements CommandContribution {
+                            newOptions: uri.parent.resolve(name).toString(),
+                            newLabel: name
+                        };
+                        this.onIUriChangedEmitter.fire(event);
```
Actually thinking more about this it would make more sense to have FileSytem.move() fire this event and have the event emitter there.
From the workspace command it looks quite odd..

hexa00 commented on packages/workspace/src/browser/workspace-frontend-module.ts:
```diff
@@ -36,8 +36,11 @@ export default new ContainerModule((bind: interfaces.Bind, unbind: interfaces.Un
     bind(MenuContribution).to(FileMenuContribution).inSingletonScope();
 
+
+
```
remove

hexa00 commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -64,6 +81,15 @@ export class WorkspaceCommandContribution implements CommandContribution {
+
+    protected readonly onIUriChangedEmitter = new Emitter<IUriChangedEvent>();
+
+    get onIUriChanged(): Event<IUriChangedEvent> {
```
We should move this to FileSytem and call it FileMovedEvent and it should be about what moved.

hexa00 replied:
Once the EditorManager receives it it can check if it has a editor for that URI and update it.

hexa00 commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -64,10 +67,14 @@ export class EditorManagerImpl implements EditorManager, WidgetFactory {
         this.currentObserver = new EditorManagerImpl.Observer('current', app);
         this.activeObserver = new EditorManagerImpl.Observer('active', app);
+        uiEvent.onIUriChanged(event => {
+            this.renameWidget(event.factoryId.toString(), event.options, event.newOptions, event.newLabel);
```
We should get the FileMovedEvent from FS here.
And then figureout the proper way to rename based on the URI change. we have all the info needed.

marcdumais-work commented:
This PR is not being actively worked-on - closing for now.
